src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -235,6 +267,70 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
+        internal static void SetInstanceAnnotations(object resource,string annotationPropertyName, ODataDeserializerProvider deserializerProvider,
+            ODataDeserializerContext readContext,ODataResource oDataResource, Dictionary<string, IDictionary<string, object>> instAnnotationDictionary)
+        {
+            IDelta delta = resource as IDelta;
+           
+            if(oDataResource.InstanceAnnotations != null && oDataResource.InstanceAnnotations.Count > 0)
+            {
+                string annotationKey = string.Empty;
+
+                if (!instAnnotationDictionary.ContainsKey(annotationKey))
+                {
+                    instAnnotationDictionary.Add(annotationKey, new Dictionary<string, object>());
+                }
+
+                AddAnnotationsToDictionary(deserializerProvider, readContext, instAnnotationDictionary, oDataResource.InstanceAnnotations, annotationKey);
+            }
+
+            foreach(ODataProperty property in oDataResource.Properties)
+            {
+                if(property.InstanceAnnotations != null && property.InstanceAnnotations.Count > 0)
+                {
+                    string annotationKey = property.Name;
+
+                    if (!instAnnotationDictionary.ContainsKey(annotationKey))
+                    {
+                        instAnnotationDictionary.Add(annotationKey, new Dictionary<string, object>());
+                    }
+
+                    AddAnnotationsToDictionary(deserializerProvider, readContext, instAnnotationDictionary, property.InstanceAnnotations, annotationKey);
+                }
+            }
+
+            if (delta == null)
+            {
+                resource.GetType().GetProperty(annotationPropertyName).SetValue(resource, instAnnotationDictionary, index: null);
+            }
+            else
+            {
+                delta.TrySetPropertyValue(annotationPropertyName, instAnnotationDictionary);
+            }                          
```

Comment:
Can you just call SetProperty() rather than duplicating that code?

Reply:
#Resolved